### ISSUE TITLE
Pb with zero?

### ISSUE THREAD

**Issue by jbcaillau** [control-toolbox/CTDirect.jl#190]
@PierreMartinon 
```julia
@def ocp2 begin
    s ∈ [0, 1], time
    y = (x, tf) ∈ R², state
    u ∈ R, control
    dx = tf(s)^2 * u(s)
    dtf = 0 * u(s) # 0
    ẏ(s) == [dx, dtf]
    x(0) == 0
    x(1) == 1
    tf(1) + 0.5∫(tf(s) * u(s)^2) → min
end
```

**Comment by PierreMartinon** [comment]
More generally, same error if you put a constant instead of 0.

**Comment by jbcaillau** [comment]
> More generally, same error if you put a constant instead of 0. Maybe this is AD related ?
> 
> Not really a problem for us since we would typically use a variable instead of a constant state, but it is a bit annoying.

@PierreMartinon yes, the error seems to come from ForwardDiff. (Actually wasted my early morning on it 🙁.) Try to investigate a bit, maybe worth sending an issue to them once the problem is clearly located.

**Comment by jbcaillau** [comment]
@amontoison does the error below ring a bell? Seems to happen when calling `ForwardDiff` on functions involving constants, as in
```julia
f(x, u) = [x[1]^2 * u, 0] # crashes while [x[1]^2 * u, 0 * u] is fine
```

```julia
ERROR: Cannot determine ordering of Dual tags ForwardDiff.Tag{ReverseDiff.var"#130#133"{typeof(+), ForwardDiff.Dual{ForwardDiff.Tag{ADNLPModels.var"#ψ#73"{CTDirect.var"#37#39"{CTDirect.DOCP}}, Float64}, Float64, 1}}, ForwardDiff.Dual{ForwardDiff.Tag{ADNLPModels.var"#ψ#73"{CTDirect.var"#37#39"{CTDirect.DOCP}}, Float64}, Float64, 1}} and ForwardDiff.Tag{ADNLPModels.var"#ψ#73"{CTDirect.var"#37#39"{CTDirect.DOCP}}, Float64}
```

**Comment by amontoison** [comment]
The issue is the combination of `ForwardDiff.jl` with `ReverseDiff.jl` for the sparse Hessian.
Can you try to create the ADNLPModel without `backend = :optimized` to confirm it?
By default, ADNLPModels.jl will do ForwardDiff.jl over ForwardDiff.jl for the Hessian (slower but more robust).

I would like to switch to DI.jl and only rely on Enzyme.jl in the future. Enzyme.jl supports both reverse and forward.

**Comment by gdalle** [comment]
I assume either ADNLPModels or CTBase must be creating some custom tags somewhere which cause this issue?
If you want to keep abreast of current events, the switch to DI in ADNLPModels is being discussed in https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/pull/302, but it's gonna be painful

**Comment by jbcaillau** [comment]
NB. Check https://github.com/control-toolbox/CTDirect.jl/issues/189#issuecomment-2270845008 for the complete example (and converging solve!)
